File: _labs/lab3/R/app/app-api.R
library(shiny)
library(bslib)
library(bsicons)
library(httr2)

api_url <- "http://127.0.0.1:8080/predict"

ui <- page_sidebar(
  title = "Penguin Mass Predictor",
  theme = bs_theme(bootswatch = "sketchy"),
  sidebar = sidebar(
    sliderInput(inputId = "bill_length", 
      label = "Bill Length (mm)",
      min = 30, 
      max = 60, 
      value = 45, 
      step = 1),
    selectInput(inputId = "sex", 
      label = "Sex", 
      choices = c("Male", "Female"), 
      selected = "Male"),
    selectInput(
      inputId = "species",
      label = "Species",
      choices = c("Adelie", "Chinstrap", "Gentoo"),
      selected = "Adelie"),
    actionButton(
      inputId = "predict", 
      label = "Predict", 
      class = "btn-primary")
  ),
  
  layout_columns(
    card(
      card_header("Penguin Parameters"),
      card_body(
      verbatimTextOutput(outputId = "vals")
      )
    ),
    card(
      card_header("Predicted Mass"),
      card_body(
        value_box(
          showcase_layout = "left center",
          title = "Grams",
          value = textOutput(outputId = "pred"),
          showcase = bs_icon("graph-up"),
          max_height = "200px",
          min_height = "200px",
        )
      )
    ),
    col_widths = c(7, 5)
  )
)

server <- function(input, output) {
  
  vals <- reactive({
    data.frame(
      bill_length_mm = input$bill_length,
      species = input$species,
      sex = tolower(input$sex)
    )
  })
  
  pred <- reactive({
    tryCatch({
      # loading notification
    showNotification("Predicting penguin mass...", 
                     type = "default", duration = 10)
        
    request_data <- vals()
      
    response <- httr2::request(api_url) |>
      httr2::req_method("POST") |>
      httr2::req_body_json(request_data, auto_unbox = FALSE) |>  
      httr2::req_perform() |>
      httr2::resp_body_json()
        
    # success notification
    showNotification("✅ Prediction successful!", 
                     type = "default", duration = 10)
        
    response$.pred[[1]]
        
    }, error = function(e) {
      error_msg <- conditionMessage(e)
      
      # error message
      if (grepl("Connection refused|couldn't connect", error_msg, ignore.case = TRUE)) {
        user_msg <- "API not available - is the server running on port 8080?"
      } else if (grepl("timeout|timed out", error_msg, ignore.case = TRUE)) {
        user_msg <- "Request timed out - API may be overloaded"
      } else {
        user_msg <- paste("API Error:", substr(error_msg, 1, 50))
      }
      
      # error notification
      showNotification(paste("❌", user_msg), type = "warn", duration = 10)
      
      # display error message
      paste("❌", user_msg)
      })
    }) |> 
    bindEvent(input$predict, ignoreInit = TRUE)

  output$pred <- renderText({
    prediction <- pred()
    
    # check for prediction 
    if (is.numeric(prediction)) {
      paste(round(prediction, 1), "grams")
    } else {
      # error message 
      prediction
    }
  })
  
  output$vals <- renderPrint({
    vals()
    })
  
}

shinyApp(ui = ui, server = server)

File: _labs/lab3/Python/app/app-api.py
from shiny import App, render, ui, reactive
import requests

api_url = 'http://127.0.0.1:8080/predict'

app_ui = ui.page_fluid(
    ui.h1("Penguin Mass Predictor"),
    ui.layout_columns(
        ui.card(
            ui.card_header("Inputs"),
            ui.input_slider(id="bill_length", label="Bill Length (mm)", min=30, max=60, value=45, step=0.1),
            ui.input_select(id="sex", label="Sex", choices=["Male", "Female"]),
            ui.input_select(id="species", label="Species", choices=["Adelie", "Chinstrap", "Gentoo"]),
            ui.input_action_button(id="predict", label="Predict")
        ),
        ui.card(
            ui.card_header("Results"),
            ui.h3("Input Values"),
            ui.output_text_verbatim(id="vals_out"),
            ui.h3("Species Encoding"),
            ui.output_text(id="species_debug"),
            ui.h3("API Connection"),
            ui.output_text(id="api_status"),
            ui.h3("Predicted Mass"),
            ui.output_text(id="pred_out")
        ),
        col_widths=[4, 8]
    )
)

def server(input, output, session):
    @reactive.calc
    def vals():
        # Create data in LIST format (what API expects)
        d = [{
            "bill_length_mm": int(input.bill_length()),
            "species_Chinstrap": int(input.species() == "Chinstrap"),
            "species_Gentoo": int(input.species() == "Gentoo"),
            "sex_male": int(input.sex() == "Male")
        }]
        return d
    
    @reactive.calc
    def api_health_check():
        """Check if API is responsive"""
        try:
            ping_url = 'http://127.0.0.1:8080/ping'
            r = requests.get(ping_url, timeout=5)
            if r.status_code == 200:
                return f"✅ API is running (ping: {r.json()})"
            else:
                return f"⚠️ API ping failed: {r.status_code}"
        except requests.exceptions.ConnectionError:
            return "❌ Cannot connect to API - is it running on port 8080?"
        except Exception as e:
            return f"❌ API health check failed: {str(e)}"
    
    @reactive.calc
    @reactive.event(input.predict)
    def pred():
        try:
            data_to_send = vals()
            print(f"\n=== PREDICTION REQUEST ===")
            print(f"Sending data to API: {data_to_send}")
            
            r = requests.post(api_url, json=data_to_send, timeout=30)
            print(f"HTTP Status Code: {r.status_code}")
            print(f"Raw response text: {r.text}")
            
            if r.status_code == 200:
                result = r.json()
                print(f"✅ Success! Parsed response: {result}")
                
                # Handle different possible response formats  
                if '.pred' in result:
                    prediction = result['.pred'][0]
                    return prediction
                elif 'predict' in result:
                    prediction = result['predict'][0]
                    return prediction
                else:
                    return f"Unexpected response format: {result}"
            else:
                return f"API Error {r.status_code}: {r.text}"
                
        except Exception as e:
            print(f"❌ Error: {e}")
            return f"Error: {str(e)}"

    @render.text
    def vals_out():
        data = vals()
        return f"{data}"
    
    @render.text
    def species_debug():
        species = input.species()
        chinstrap = int(species == "Chinstrap")
        gentoo = int(species == "Gentoo")
        
        if chinstrap == 0 and gentoo == 0:
            return f"Selected: {species} → Adelie (reference category)"
        elif chinstrap == 1:
            return f"Selected: {species} → Chinstrap"
        elif gentoo == 1:
            return f"Selected: {species} → Gentoo"
    
    @render.text
    def api_status():
        return api_health_check()

    @render.text
    def pred_out():
        result = pred()
        if isinstance(result, (int, float)):
            return f"{round(result, 1)} grams"
        else:
            return str(result)

app = App(app_ui, server)